### PR TITLE
chore: don't re-render functions on every keystroke

### LINKED ISSUE #1754: typing in notebooks flux editor is slow
Typing in the notebooks flux editor is slow. This causes weird behavior when you have multiple tabs open with different notebooks.

I've noticed that if I close the functions panel, the editor is faster. But others have reported that does not solve the slowness for them.

I think some component tree is being re-rendered unnecessarily with every key stroke. 

Note: we do not want to add debounce to the editor to work around this problem. We should try and reduce the number of re-renders occurring first

Slack convo: https://influxdata.slack.com/archives/C01E1H8D04Q/p1623789945202400

### PR DESCRIPTION
**Pull request by genehynson**
Closes #1754 

Functions were being rendered multiple times on every keystroke and was causing a slow typing experience. I wrapped it in `useMemo` and now it only renders when the Function's search box is used. 

### Notebooks:

Before: 1 keystroke = 2 renders = 40ms render time
After: 1 keystroke = 0 renders = 16ms render time

Screen Shot 2021-07-09 at 1 47 17 PM
Screen Shot 2021-07-09 at 1 44 25 PM

### Data Explorer:

Before: 1 keystroke = 10 renders = 47ms render time
After: 1 keystroke = 0 renders = 5ms render time

image
image

### Example typing speed improvement

Before:

https://user-images.githubusercontent.com/6411855/125136684-28560200-e0c0-11eb-8b90-ac102ba1d21c.mov

After:

https://user-images.githubusercontent.com/6411855/125136787-53d8ec80-e0c0-11eb-85c3-86c86d4d7dc5.mov

### CODE REVIEW

File: src/timeMachine/components/fluxFunctionsToolbar/FluxFunctionsToolbar.tsx [hunk 1]
```diff
@@ -56,36 +52,22 @@ class FluxFunctionsToolbar extends PureComponent<Props, State> {
...
-
-  private handleClickFunction = (func: FluxToolbarFunction) => {
-    this.props.onInsertFluxFunction(func)
-  }
-}
-
-const mstp = (state: AppState) => {
-  const activeQueryText = getActiveQuery(state).text
-
-  return {activeQueryText}
```
Comment (genehynson) [inline src/timeMachine/components/fluxFunctionsToolbar/FluxFunctionsToolbar.tsx:82]: I didn't see any reason for this component to need the `activeQueryText` so I removed that dependency

File: src/flows/pipes/RawFluxEditor/functions.tsx [hunk 2]
```diff
@@ -48,57 +48,59 @@ const Functions: FC<Props> = ({onSelect}) => {
     [search]
   )
 
-  let fnComponent
+  return useMemo(() => {
```
Comment (genehynson) [inline src/flows/pipes/RawFluxEditor/functions.tsx:51]: the diff looks crazy, but all I did was wrap a large section of this component in a `useMemo` which caused the indentation to change and then prettier adjusted some things too

File: src/timeMachine/components/fluxFunctionsToolbar/FluxFunctionsToolbar.tsx [hunk 3]
```diff
@@ -28,22 +25,21 @@ interface OwnProps {
 type ReduxProps = ConnectedProps<typeof connector>
 type Props = OwnProps & ReduxProps
 
-interface State {
-  searchTerm: string
-}
+const FluxFunctionsToolbar: FC<Props> = (props: Props) => {
```
Comment (genehynson) [inline src/timeMachine/components/fluxFunctionsToolbar/FluxFunctionsToolbar.tsx:28]: converted this component to a FC to take advantage of useMemo

**Comment by ui-ci-bot-influx** [comment]
Metric | Score | Level
------------ | ------------- | ------------- 
performance | 53 | :red_circle:
accessibility | 96 | :green_circle:
seo | 90 | :green_circle:
pwa | null | :hankey:
best-practices | 100 | :green_circle:


[Compare your scores with master branch](https://influxdb.aws.influxdata.io/orgs/844910ece80be8bc/dashboards/073c4b95f1089000?lower=now%28%29%20-%202d&vars%5Blighthouse_score_branch%5D=1959)

**Review by drdelambre — APPROVED** [review]
Looks good. That ish is expensive

File: src/timeMachine/components/fluxFunctionsToolbar/FluxFunctionsToolbar.tsx (hunk 1 above)
Comment (genehynson) [reply src/timeMachine/components/fluxFunctionsToolbar/FluxFunctionsToolbar.tsx]: ok I lied, it's needed

File: src/flows/pipes/RawFluxEditor/functions.tsx (hunk 2 above)
Comment (hoorayimhelping) [reply src/flows/pipes/RawFluxEditor/functions.tsx:51]: protip: add `?w=1` to the url to ignore whitespace

https://github.com/influxdata/ui/pull/1959/files?w=1

Comment (genehynson) [reply src/flows/pipes/RawFluxEditor/functions.tsx:51]: DUDE

**Review by Asalem1 — APPROVED** [review]

File: src/timeMachine/components/fluxFunctionsToolbar/FluxFunctionsToolbar.tsx (hunk 1 above)
Comment (genehynson) [reply src/timeMachine/components/fluxFunctionsToolbar/FluxFunctionsToolbar.tsx:82]: lied again, it's not needed. The issue was that we weren't re-rendering when props changed

**Comment by genehynson** [comment]
Turns out we set the `editorInstance` after the editor is mounted which creates a chicken and the egg problem here because the functions are rendered without the editorInstance and because they are not re-rendered, the "Inject" button doesn't work initially. So I memoized the parent components where `editorInstance` is stored. This is necessary because simply adding the `onSelect` function to the memoize dependency array causes the re-render on every keystroke. So the parent component should only re-render when specific deps are changed, such as the `editorInstance`, the `activeQueryTabIndex`, or if the user wants to hide the function panel

File: src/timeMachine/components/fluxFunctionsToolbar/FluxFunctionsToolbar.tsx [hunk 4]
```diff
@@ -56,36 +55,22 @@ class FluxFunctionsToolbar extends PureComponent<Props, State> {
...
 
 const mdtp = {
   onSetActiveQueryText: setActiveQueryText,
 }
 
-const connector = connect(mstp, mdtp)
+const connector = connect(null, mdtp)
 
-export default connector(ErrorHandling(FluxFunctionsToolbar))
+export default connector(FluxFunctionsToolbar)
```
Comment (Asalem1) [inline src/timeMachine/components/fluxFunctionsToolbar/FluxFunctionsToolbar.tsx:76]: nit: this was refactored to a functional component, why not use hooks here for dispatching the query?

**Review by Asalem1 — APPROVED** [review]
Still looks good to me

File: src/timeMachine/components/fluxFunctionsToolbar/FluxFunctionsToolbar.tsx (hunk 4 above)
Comment (genehynson) [reply src/timeMachine/components/fluxFunctionsToolbar/FluxFunctionsToolbar.tsx]: turns out there actually is no need for this function to be connected to redux at all so I just removed it

**Pull request merged**